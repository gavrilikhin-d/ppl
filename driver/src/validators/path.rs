use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
