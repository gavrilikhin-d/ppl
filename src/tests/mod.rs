use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
