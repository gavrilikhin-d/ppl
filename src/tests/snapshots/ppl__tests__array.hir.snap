---
source: src/tests/mod.rs
expression: hir
---
let mut arr: Array<Integer> = `<:Type<Integer>> [ ]`((Type<Integer>:Type<Integer>))
`println <:Array<Integer>>`((arr:Array<Integer>))
if `<:Reference<Array<Integer>>> is empty`((&arr:ReferenceMut<Array<Integer>>)):
	`println <:String>`("Empty")

`push <:Integer> to <:ReferenceMut<Array<Integer>>>`(1, (&arr:ReferenceMut<Array<Integer>>))
`println <:Array<Integer>>`((arr:Array<Integer>))
if `<:Reference<Array<Integer>>> is not empty`((&arr:ReferenceMut<Array<Integer>>)):
	`println <:String>`("Not empty")

`println <:Integer>`(`clone <:Reference<Integer>>`((*`<:ReferenceMut<Array<Integer>>> [ <:Integer> ]`((&arr:ReferenceMut<Array<Integer>>), 0):Integer)))
let mut i: Integer = 2
while `<:Integer> <= <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), 10):
	`push <:Integer> to <:ReferenceMut<Array<Integer>>>`(`clone <:Reference<Integer>>`((i:Integer)), (&arr:ReferenceMut<Array<Integer>>))
	`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)
	`println <:Array<Integer>>`((arr:Array<Integer>))

`destroy <:ReferenceMut<Integer>>`((i:Integer))

==MONOMORPHIZED==


fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


fn<Integer> allocate <n: Integer> <$arg1: Type<Integer>> -> MemoryAddress:
	let $tmp@825: MemoryAddress = `allocate <:Integer> bytes`(`<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((n:Integer)), `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))))
	`destroy <:ReferenceMut<Integer>>`((n:Integer))
	return ($tmp@825:MemoryAddress)


fn<Integer> <$arg0: Type<Integer>> [ ] -> Array<Integer>:
	let capacity: Integer = 8
	let data: MemoryAddress = `allocate <:Integer> <:Type<Integer>>`(`clone <:Reference<Integer>>`((capacity:Integer)), (Type<Integer>:Type<Integer>))
	let $tmp@427: Array<Integer> = Array<Integer> { size: 0, capacity: `clone <:Reference<Integer>>`((capacity:Integer)), data: (data:MemoryAddress) }
	`destroy <:ReferenceMut<Integer>>`((capacity:Integer))
	return ($tmp@427:Array<Integer>)


fn<Integer> String from <array: Array<Integer>> -> String:
	let mut str: String = "["
	if `<:Reference<Array<Integer>>> is not empty`((&array:Reference<Array<Integer>>)):
		`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), 0):Integer)))
		let mut i: Integer = 1
		while `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((array:Array<Integer>).size)):
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), ", ")
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), `clone <:Reference<Integer>>`((i:Integer))):Integer)))
			`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)

		`destroy <:ReferenceMut<Integer>>`((i:Integer))

	`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), "]")
	let $tmp@1936: String = `clone <:Reference<String>>`((str:String))
	`destroy <:ReferenceMut<String>>`((str:String))
	return ($tmp@1936:String)


fn println <x: Array<Integer>> -> None:
	let $tmp@176: None = `println <:String>`(`String from <:Array<Integer>>`((x:Array<Integer>)))
	return ($tmp@176:None)



fn<Integer> <array: Reference<Array<Integer>>> is empty -> Bool:
	let $tmp@1612: Bool = `<:Integer> == <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), 0)
	return ($tmp@1612:Bool)


fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


fn<Integer> allocate <n: Integer> <$arg1: Type<Integer>> -> MemoryAddress:
	let $tmp@825: MemoryAddress = `allocate <:Integer> bytes`(`<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((n:Integer)), `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))))
	`destroy <:ReferenceMut<Integer>>`((n:Integer))
	return ($tmp@825:MemoryAddress)


@mangle_as("address_of")
fn<Integer> address of <ref: Reference<Integer>> -> MemoryAddress

fn<Integer> push <x: Integer> to <array: ReferenceMut<Array<Integer>>> -> None:
	let bytes: Integer = `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))
	if `<:Integer> == <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), `clone <:Reference<Integer>>`((*array:Array<Integer>).capacity)):
		let new_capacity: Integer = `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).capacity), 2)
		let new_data: MemoryAddress = `allocate <:Integer> <:Type<Integer>>`(`clone <:Reference<Integer>>`((new_capacity:Integer)), (Type<Integer>:Type<Integer>))
		let mut i: Integer = 0
		while `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((*array:Array<Integer>).size)):
			let offset: Integer = `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((bytes:Integer)))
			let $tmp@1353: MemoryAddress = `<:MemoryAddress> + <:Integer>`((*array:Array<Integer>).data, `clone <:Reference<Integer>>`((offset:Integer)))
			let $tmp@1378: MemoryAddress = `<:MemoryAddress> + <:Integer>`((new_data:MemoryAddress), `clone <:Reference<Integer>>`((offset:Integer)))
			`copy <:Reference<Integer>> bytes from <:Reference<MemoryAddress>> to <:Reference<MemoryAddress>>`((&bytes:Reference<Integer>), (&$tmp@1353:Reference<MemoryAddress>), (&$tmp@1378:Reference<MemoryAddress>))
			`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)
			`destroy <:ReferenceMut<Integer>>`((offset:Integer))

		`free <:Reference<MemoryAddress>>`((&(*array:Array<Integer>).data:ReferenceMut<MemoryAddress>))
		(*array:Array<Integer>).data = (new_data:MemoryAddress)
		`destroy <:ReferenceMut<Integer>>`((*array:Array<Integer>).capacity)
		(*array:Array<Integer>).capacity = `clone <:Reference<Integer>>`((new_capacity:Integer))
		`destroy <:ReferenceMut<Integer>>`((new_capacity:Integer))
		`destroy <:ReferenceMut<Integer>>`((i:Integer))

	let $tmp@1506: MemoryAddress = `address of <:Reference<Integer>>`((&x:Reference<Integer>))
	let $tmp@1524: MemoryAddress = `<:MemoryAddress> + <:Integer>`((*array:Array<Integer>).data, `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), `clone <:Reference<Integer>>`((bytes:Integer))))
	`copy <:Reference<Integer>> bytes from <:Reference<MemoryAddress>> to <:Reference<MemoryAddress>>`((&bytes:Reference<Integer>), (&$tmp@1506:Reference<MemoryAddress>), (&$tmp@1524:Reference<MemoryAddress>))
	`<:ReferenceMut<Integer>> += <:Integer>`((&(*array:Array<Integer>).size:ReferenceMut<Integer>), 1)
	`destroy <:ReferenceMut<Integer>>`((bytes:Integer))


fn<Integer> String from <array: Array<Integer>> -> String:
	let mut str: String = "["
	if `<:Reference<Array<Integer>>> is not empty`((&array:Reference<Array<Integer>>)):
		`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), 0):Integer)))
		let mut i: Integer = 1
		while `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((array:Array<Integer>).size)):
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), ", ")
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), `clone <:Reference<Integer>>`((i:Integer))):Integer)))
			`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)

		`destroy <:ReferenceMut<Integer>>`((i:Integer))

	`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), "]")
	let $tmp@1936: String = `clone <:Reference<String>>`((str:String))
	`destroy <:ReferenceMut<String>>`((str:String))
	return ($tmp@1936:String)


fn println <x: Array<Integer>> -> None:
	let $tmp@176: None = `println <:String>`(`String from <:Array<Integer>>`((x:Array<Integer>)))
	return ($tmp@176:None)



fn<Integer> <array: Reference<Array<Integer>>> is not empty -> Bool:
	let $tmp@1669: Bool = `<:Integer> > <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), 0)
	return ($tmp@1669:Bool)


fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


@mangle_as("read_memory")
fn<Integer> <ty: Type<Integer>> at <address: Reference<MemoryAddress>> -> ReferenceMut<Integer>

@mangle_as("read_memory")
fn<Integer> <ty: Type<Integer>> at <address: Reference<MemoryAddress>> -> ReferenceMut<Integer>

fn<Integer> <array: ReferenceMut<Array<Integer>>> [ <i: Integer> ] -> ReferenceMut<Integer>:
	let $tmp@784: String = "Index out of bounds"
	`assert <:Bool> <:Reference<String>>`(`<:Bool> and <:Bool>`(`<:Integer> <= <:Integer>`(0, `clone <:Reference<Integer>>`((i:Integer))), `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((*array:Array<Integer>).size))), (&$tmp@784:Reference<String>))
	let address: MemoryAddress = `<:MemoryAddress> + <:Integer>`((*array:Array<Integer>).data, `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))))
	let mut $tmp@858: ReferenceMut<Integer> = `<:Type<Integer>> at <:Reference<MemoryAddress>>`((Type<Integer>:Type<Integer>), (&address:Reference<MemoryAddress>))
	`destroy <:ReferenceMut<Integer>>`((i:Integer))
	`destroy <:ReferenceMut<String>>`(($tmp@784:String))
	return ($tmp@858:ReferenceMut<Integer>)


@mangle_as("integer_as_string")
fn String from <$arg0: Integer> -> String

fn println <x: Integer> -> None:
	let $tmp@176: None = `println <:String>`(`String from <:Integer>`((x:Integer)))
	return ($tmp@176:None)



@mangle_as("integer_less_integer")
fn <$arg0: Integer> < <$arg1: Integer> -> Bool

fn <x: Integer> > <y: Integer> -> Bool:
	let $tmp@1032: Bool = `<:Integer> < <:Integer>`((y:Integer), (x:Integer))
	return ($tmp@1032:Bool)



fn <x: Integer> <= <y: Integer> -> Bool:
	let $tmp@1068: Bool = `not <:Bool>`(`<:Integer> > <:Integer>`((x:Integer), (y:Integer)))
	return ($tmp@1068:Bool)



fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


fn<Integer> size of <ty: Type<Integer>> -> Integer:
	let $tmp@4582: Integer = `clone <:Reference<Integer>>`((ty:Type<Integer>).size)
	return ($tmp@4582:Integer)


fn<Integer> allocate <n: Integer> <$arg1: Type<Integer>> -> MemoryAddress:
	let $tmp@825: MemoryAddress = `allocate <:Integer> bytes`(`<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((n:Integer)), `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))))
	`destroy <:ReferenceMut<Integer>>`((n:Integer))
	return ($tmp@825:MemoryAddress)


@mangle_as("address_of")
fn<Integer> address of <ref: Reference<Integer>> -> MemoryAddress

fn<Integer> push <x: Integer> to <array: ReferenceMut<Array<Integer>>> -> None:
	let bytes: Integer = `size of <:Type<Integer>>`((Type<Integer>:Type<Integer>))
	if `<:Integer> == <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), `clone <:Reference<Integer>>`((*array:Array<Integer>).capacity)):
		let new_capacity: Integer = `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).capacity), 2)
		let new_data: MemoryAddress = `allocate <:Integer> <:Type<Integer>>`(`clone <:Reference<Integer>>`((new_capacity:Integer)), (Type<Integer>:Type<Integer>))
		let mut i: Integer = 0
		while `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((*array:Array<Integer>).size)):
			let offset: Integer = `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((bytes:Integer)))
			let $tmp@1353: MemoryAddress = `<:MemoryAddress> + <:Integer>`((*array:Array<Integer>).data, `clone <:Reference<Integer>>`((offset:Integer)))
			let $tmp@1378: MemoryAddress = `<:MemoryAddress> + <:Integer>`((new_data:MemoryAddress), `clone <:Reference<Integer>>`((offset:Integer)))
			`copy <:Reference<Integer>> bytes from <:Reference<MemoryAddress>> to <:Reference<MemoryAddress>>`((&bytes:Reference<Integer>), (&$tmp@1353:Reference<MemoryAddress>), (&$tmp@1378:Reference<MemoryAddress>))
			`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)
			`destroy <:ReferenceMut<Integer>>`((offset:Integer))

		`free <:Reference<MemoryAddress>>`((&(*array:Array<Integer>).data:ReferenceMut<MemoryAddress>))
		(*array:Array<Integer>).data = (new_data:MemoryAddress)
		`destroy <:ReferenceMut<Integer>>`((*array:Array<Integer>).capacity)
		(*array:Array<Integer>).capacity = `clone <:Reference<Integer>>`((new_capacity:Integer))
		`destroy <:ReferenceMut<Integer>>`((new_capacity:Integer))
		`destroy <:ReferenceMut<Integer>>`((i:Integer))

	let $tmp@1506: MemoryAddress = `address of <:Reference<Integer>>`((&x:Reference<Integer>))
	let $tmp@1524: MemoryAddress = `<:MemoryAddress> + <:Integer>`((*array:Array<Integer>).data, `<:Integer> * <:Integer>`(`clone <:Reference<Integer>>`((*array:Array<Integer>).size), `clone <:Reference<Integer>>`((bytes:Integer))))
	`copy <:Reference<Integer>> bytes from <:Reference<MemoryAddress>> to <:Reference<MemoryAddress>>`((&bytes:Reference<Integer>), (&$tmp@1506:Reference<MemoryAddress>), (&$tmp@1524:Reference<MemoryAddress>))
	`<:ReferenceMut<Integer>> += <:Integer>`((&(*array:Array<Integer>).size:ReferenceMut<Integer>), 1)
	`destroy <:ReferenceMut<Integer>>`((bytes:Integer))


@mangle_as("integer_plus_integer")
fn <$arg0: Integer> + <$arg1: Integer> -> Integer

fn <self: ReferenceMut<Integer>> += <other: Integer> -> None:
	(self:ReferenceMut<Integer>) = `<:Integer> + <:Integer>`((*self:Integer), (other:Integer))


fn<Integer> String from <array: Array<Integer>> -> String:
	let mut str: String = "["
	if `<:Reference<Array<Integer>>> is not empty`((&array:Reference<Array<Integer>>)):
		`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), 0):Integer)))
		let mut i: Integer = 1
		while `<:Integer> < <:Integer>`(`clone <:Reference<Integer>>`((i:Integer)), `clone <:Reference<Integer>>`((array:Array<Integer>).size)):
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), ", ")
			`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), `String from <:Integer>`((*`<:Reference<Array<Integer>>> [ <:Integer> ]`((&array:Reference<Array<Integer>>), `clone <:Reference<Integer>>`((i:Integer))):Integer)))
			`<:ReferenceMut<Integer>> += <:Integer>`((&i:ReferenceMut<Integer>), 1)

		`destroy <:ReferenceMut<Integer>>`((i:Integer))

	`<:ReferenceMut<String>> += <:String>`((&str:ReferenceMut<String>), "]")
	let $tmp@1936: String = `clone <:Reference<String>>`((str:String))
	`destroy <:ReferenceMut<String>>`((str:String))
	return ($tmp@1936:String)


fn println <x: Array<Integer>> -> None:
	let $tmp@176: None = `println <:String>`(`String from <:Array<Integer>>`((x:Array<Integer>)))
	return ($tmp@176:None)
